Move to Microsoft.Extensions.Logging (squashed/rebased #899)

gbirchmeier opened the pull request:
Rebased PR #899 on master.  (That PR addresses #679.)

I squashed all 27 commits from #899 together into one commit to ease the rebase process.  (Some of the later commits undid or reworked the earlier commits which made the incremental-commit rebasing kind of hairy.)


--  
close #899 - PR superseded by this one
close #679 - source issue

gbirchmeier commented:
@jkulubya Maybe you want to scan over my update, especially ThreadedSocketAcceptor which had changed quite a bit since your PR was first submitted.  For that file, I ended up reverting it to the current master an then making new changes as appropriate.

gbirchmeier commented:
@jkulubya Is there a reason that you often used e.g. `somelogger.Log(LogLevel.Error, "blah")` instead of `somelogger.LogError("blah")`?

gbirchmeier commented:
A couple vital things that I can see we need right off the bat:

* **A file-based logger impl:** a file-based message log is the number-one thing that I always look for when investigating issues.  Any release of QF/n that includes this fix *****MUST***** ship with a file logger that is easy to set up.  (It completely blows my mind that MEL doesn't provide a default File-based logger implementation)

* **How-to Documentation:** We'll need a doc page that shows how to set up a logger with the new system, with a cheatsheet for people who don't know MEL (like me!).  I see that all the examples now set up a console logger, and that's nice, but... back to the first bullet -- I want a file logger!

My first task is to actually learn MEL, as I will not be confident moving forward with this until I understand it on a deep enough level.  I have questions about some of the class designs in this PR, and I'm sure they will start to make sense after I slog through some MEL docs and create some demo apps.

jkulubya commented:
> @jkulubya Is there a reason that you often used e.g. `somelogger.Log(LogLevel.Error, "blah")` instead of `somelogger.LogError("blah")`?

There's no functional difference between the two, but I found it easier to ctrl+F with `somelogger.Log()`

jkulubya commented:
> @jkulubya Maybe you want to scan over my update, especially ThreadedSocketAcceptor which had changed quite a bit since your PR was first submitted. For that file, I ended up reverting it to the current master an then making new changes as appropriate.

Looks fine to me

jkulubya commented:
> A couple vital things that I can see we need right off the bat:
> 
>     * **A file-based logger impl:** a file-based message log is the number-one thing that I always look for when investigating issues.  Any release of QF/n that includes this fix _**MUST**_ ship with a file logger that is easy to set up.  (It completely blows my mind that MEL doesn't provide a default File-based logger implementation)

Yeah, it's a little unfortunate that there's no in-built option for logging to files, but they closed the door pretty firmly on that. I did write a sample for serilog in `/examples/Standalone/SerilogLog`. The config for that (in `Program.cs`) is slightly more complicated than you'd typically write only to match the old behaviour, and you end up with a file system that looks like this if you roll by the minute:

```
$ ls
FIX.4.0-CLIENT1-EXECUTOR.event202501090846.log     FIX.4.4-CLIENT1-EXECUTOR.event202501090846.log     FIX.4.4-CLIENT2-EXECUTOR.event202501090846.log     Non-Session-Log.event.current.log
FIX.4.0-CLIENT1-EXECUTOR.messages202501090846.log  FIX.4.4-CLIENT1-EXECUTOR.messages202501090846.log  FIX.4.4-CLIENT2-EXECUTOR.messages202501090846.log
FIX.4.0-CLIENT1-EXECUTOR.messages202501090847.log  FIX.4.4-CLIENT1-EXECUTOR.messages202501090847.log  FIX.4.4-CLIENT2-EXECUTOR.messages202501090847.log

$ cat FIX.4.0-CLIENT1-EXECUTOR.event202501090846.log
Created session
Connecting to 127.0.0.1 on port 5001
Connection succeeded
Session reset: ResetOnLogon
Initiated logon request
Received logon

$ cat FIX.4.0-CLIENT1-EXECUTOR.messages202501090846.log 
8=FIX.4.09=6635=A34=149=CLIENT152=20250109-13:46:4556=EXECUTOR98=0108=1010=151
8=FIX.4.09=6635=A34=149=EXECUTOR52=20250109-13:46:4556=CLIENT198=0108=1010=151
8=FIX.4.09=5435=034=249=EXECUTOR52=20250109-13:46:5556=CLIENT110=110
8=FIX.4.09=5435=034=249=CLIENT152=20250109-13:46:5556=EXECUTOR10=110
```   

> 
>     * **How-to Documentation:** We'll need a doc page that shows how to set up a logger with the new system, with a cheatsheet for people who don't know MEL (like me!).  I see that all the examples now set up a console logger, and that's nice, but... that example is not enough.  There needs to be a page that shows users how to quickly configure a console log OR a file log, and gives pointers on what else they can do.
> 
> 
> My first task is to actually learn MEL myself, as I will not be confident moving forward with this until I understand it on a deep enough level. I have questions about some of the class designs in this PR, and I'm sure they will start to make sense after I slog through some MEL docs and create some demo apps.

I can start a write up based on the Serilog example and we can go from there on the docs

gbirchmeier commented on QuickFIXn/Logger/LogFactoryAdapter.cs:
```diff
@@ -0,0 +1,66 @@
+    internal LogFactoryAdapter(ILogFactory logFactory)
+    {
+        _logFactory = logFactory;
+        _nonSessionLogger = new Lazy<ILogger>(() => new LogAdapter(_logFactory.CreateNonSessionLog()));
```
TIL about `Lazy` for the first time 🙂

gbirchmeier commented on QuickFIXn/SessionState.cs:
```diff
@@ -402,7 +402,6 @@ protected virtual void Dispose(bool disposing)
             if (_disposed) return;
             if (disposing)
             {
-                Log.Dispose();
```
@jkulubya What if `Log` is (e.g.) a `LogAdapter`, which is an `IDisposable`?

Although... SessionState doesn't create the Log object, so now I think it should never have been disposing it.  The containing Session object creates the log, and should probably be doing that disposal instead.

Is that anything you looked into or thought about while working on this?

gbirchmeier commented on QuickFIXn/ThreadedSocketAcceptor.cs:
```diff
@@ -32,18 +34,56 @@ public class ThreadedSocketAcceptor : IAcceptor
+                // must save a ref to it so we can dispose it later. Any other loggerFactory is owned by someone else
+                // so we'll leave the dispose up to them. This should be removed eventually together with the old ILog
+                // and ILogFactory.
+                _logFactoryAdapter = lfa;
```
@jkulubya Looks like you forgot to add this to `Dispose()`, right?  I'll do it in my current branch.

gbirchmeier replied:
@jkulubya Update: I see you did have a dispose call, but you put it at the end of `Stop()`.  This is an error, because Start/Stop can be called on this object multiple times.  ~Dispose()~ is the right place for it, and I made that change.

gbirchmeier commented on QuickFIXn/Session.cs:
```diff
@@ -362,7 +363,19 @@ public bool Send(string message)
+                const LogLevel messagesLogLevel = LogLevel.Information;
+                if (Log.IsEnabled(messagesLogLevel))
+                {
+                    using (Log.BeginScope(new Dictionary<string, object>
```
@jkulubya  What is this scope doing?  My brain is refusing to understand what this code does.

josh-shaw-dev replied:
BeginScope adds the parameters from the dictionary for all logging done inside that using scope

So in serilog land for json sink you would have what ever props are in the dictionary with the log so in this case 
Properties would be this for a fix44 AllocationReport message

{
  "Message": "The actual message..."
  "Properties" : {"MessageType": "AllocationReport"}
}


Azure App insights would be similar though would end up in custom dimensions I think its called

gbirchmeier commented on QuickFIXn/Session.cs:
```diff
@@ -630,19 +660,19 @@ internal void Next(MessageBuilder msgBuilder)
             catch (FieldNotFoundException e)
             {
-                Log.OnEvent("Rejecting invalid message, field not found: " + e.Message);
+                Log.Log(LogLevel.Information, e, "Rejecting invalid message, field not found: {Message}", e.Message);
```
Is the `{Message}` part redundant?

gbirchmeier replied:
These probably are, but I'm going to leave them in anyway (and change more log calls to be like this), because it's nice to have the actual error message in the first line.

After seeing this "in the wild" for a bit, I may come reevaluate this decision.

gbirchmeier commented on QuickFIXn/Logger/LogEventIds.cs:
```diff
@@ -0,0 +1,9 @@
+
+internal static class LogEventIds
+{
+    internal static readonly EventId IncomingMessage = 7702;
```
@jkulubya Is there specific reason you decided on these id values, or was it just an arbitrary choice?

gbirchmeier replied:
@jkulubya Your version had IDs 7702 and 7203 (not 7_7_03), which I'm pretty sure was a typo.  I changed them to 7702/7703.

Let me know if this was your intent.  Also, can you advise how you chose these numbers initially?

josh-shaw-dev commented on QuickFIXn/SocketReader.cs:
```diff
@@ -19,7 +20,15 @@ public class SocketReader : IDisposable
+    {
+        get
+        {
+            if (_qfSession?.Log is { } logger) return logger;
```
I agree with the syntax is terrible for readability, 

You could easily just use ILogger instead of { }
`if (_qfSession?.Log is ILogger logger) return logger;`

Or is not null is another option
`if (_qfSession?.Log is not null) return _qfSession.Log;`

gbirchmeier replied:
Changed the whole property to:

`private ILogger UnconditionalLogger => _qfSession?.Log is not null ? _qfSession.Log : _nonSessionLog;`

Way better.

josh-shaw-dev commented on QuickFIXn/Session.cs:
```diff
@@ -619,7 +649,7 @@ internal void Next(MessageBuilder msgBuilder)
             {
                 if (e.InnerException is not null)
-                    Log.OnEvent(e.InnerException.Message);
+                    Log.Log(LogLevel.Error, "{Message}", e.InnerException.Message);
```
Is there a reason why to not include the exception for this one when its an error level so you would get a stack trace etc when you have the ones below including them

gbirchmeier replied:
one more thing: Such an exception doesn't actually indicate any error in the running app.  Sometimes a badly-implemented counterparty can send bad messages, and your app just has to deal.  And the correct way to deal is to 35=3 reject.

gbirchmeier commented:
I've drafted a documentation page.  This would not go live until after v1.14 is released.  Here are screenshots.  Comments welcome.

<img width="946" height="752" alt="Screenshot 2025-07-28 at 4 39 16 PM" src="https://github.com/user-attachments/assets/6a2f86a8-827e-46e4-a933-d5b904170d88" />

<img width="944" height="704" alt="Screenshot 2025-07-28 at 4 39 26 PM" src="https://github.com/user-attachments/assets/e9136a3b-4172-46c5-b2c5-5e32ef1cb037" />

<img width="943" height="737" alt="Screenshot 2025-07-28 at 4 39 37 PM" src="https://github.com/user-attachments/assets/1cebf8d9-83d1-418d-8996-c4f9c89c8eec" />

jkulubya commented:
Looks good, I’d add a note saying that QF now depends on MEL.Abstractions v6, which should be compatible with (on the application developer side) all MEL and dotnet versions from 6 to ? (for the foreseeable future).On Jul 28, 2025, at 5:41 PM, Grant Birchmeier ***@***.***> wrote:﻿gbirchmeier left a comment (connamara/quickfixn#971)
I've drafted a documentation page.  This would not go live until after v1.14 is released.  Here are screenshots.  Comments welcome.
Screenshot.2025-07-28.at.4.39.16.PM.png (view on web)
Screenshot.2025-07-28.at.4.39.26.PM.png (view on web)
Screenshot.2025-07-28.at.4.39.37.PM.png (view on web)

—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you were mentioned.Message ID: ***@***.***>

gbirchmeier commented:
Thanks.  I'm assuming you are reading them from emails, as I didn't leave them up for long because I decided they needed more work.  I will post updated versions today or tomorrow, and have this PR in "final review" form.

gbirchmeier commented:
@jkulubya  Here is my final version of the docs.  The "Distinguishing Different Categories of Logs" is the big thing that I added since my last attempt; as I was going over the Serilog config I realized that this was super-important info.  (Also, I simplified the Serilog config as you'll surely see -- Source and EventId is really all we need.

<img width="904" height="672" alt="Screenshot 2025-08-08 at 11 19 41 AM" src="https://github.com/user-attachments/assets/ec11484f-27a3-4ea2-b724-fbca351bccf8" />
<img width="907" height="623" alt="Screenshot 2025-08-08 at 11 19 58 AM" src="https://github.com/user-attachments/assets/7e68a36d-1a21-4c0d-a652-4ec4482f6d28" />
<img width="901" height="914" alt="Screenshot 2025-08-08 at 11 20 11 AM" src="https://github.com/user-attachments/assets/7d3b3904-9aa5-4a5c-ab77-3ab0b7186d7a" />

gbirchmeier commented:
Ok, I think this is code complete.  I made a lot of small revisions which you can see, but the 2 most significant are:

* Don't obsolete the old interface
* Move Serilog back to a standalone solution.  I don't want the main QF/n solution to "need" to install Serilog.

I welcome reviews of my changes.  I'll be away next week, but the week after I'd like to merge this thing.

Rob-Hague commented on Examples/Executor/Examples.Executor.csproj:
```diff
@@ -40,4 +40,9 @@
   </ItemGroup>
 
+  <ItemGroup>
+    <PackageReference Include="Microsoft.Extensions.Logging" Version="8.0.1" />
+    <PackageReference Include="Microsoft.Extensions.Logging.Console" Version="8.0.1" />
```
Since the second depends on the first, it could just be

```suggestion
    <PackageReference Include="Microsoft.Extensions.Logging.Console" Version="8.0.1" />
```

(same for the other examples)

Rob-Hague commented on QuickFIXn/QuickFix.csproj:
```diff
@@ -43,4 +43,8 @@
   </ItemGroup>
+
+  <ItemGroup>
+    <PackageReference Include="Microsoft.Extensions.Logging.Abstractions" Version="6.0.0" />
```
So I initially argued to use 6.0.0 because at the time 8.0.2 was pulling in DiagnosticSource unnecessarily. They have since fixed that in 8.0.3, so if it were up to me I would use 8.0.3, but ultimately it does not matter since it is mainly just interface definitions

https://www.nuget.org/packages/Microsoft.Extensions.Logging.Abstractions/6.0.0#dependencies-body-tab
https://www.nuget.org/packages/Microsoft.Extensions.Logging.Abstractions/8.0.2#dependencies-body-tab
https://www.nuget.org/packages/Microsoft.Extensions.Logging.Abstractions/8.0.3#dependencies-body-tab

Rob-Hague commented on QuickFIXn/SocketReader.cs:
```diff
@@ -19,7 +20,9 @@ public class SocketReader : IDisposable
-    private readonly NonSessionLog _nonSessionLog;
+    private readonly ILogger _nonSessionLog;
+
+    private ILogger UnconditionalLogger => _qfSession?.Log is not null ? _qfSession.Log : _nonSessionLog;
```
another syntax choice 🙂 

```suggestion
    private ILogger UnconditionalLogger => _qfSession?.Log ?? _nonSessionLog;
```

gbirchmeier replied:
of course, duhhhhhhhhhh on me!  Will be in the next commit.

Rob-Hague reviewed (approved):
Nice job both

There are a few instances where a class is made internal, or a constructor is made internal so the type is no longer publicly instantiable. I assume these are intended, just highlighting to double check

jkulubya commented:
> @jkulubya Here is my final version of the docs. The "Distinguishing Different Categories of Logs" is the big thing that I added since my last attempt; as I was going over the Serilog config I realized that this was super-important info. (Also, I simplified the Serilog config as you'll surely see -- Source and EventId is really all we need.
> <img alt="Screenshot 2025-08-08 at 11 19 41 AM" width="904" height="672" src="https://private-user-images.githubusercontent.com/237288/476114173-ec11484f-27a3-4ea2-b724-fbca351bccf8.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTQ5NTIzNTgsIm5iZiI6MTc1NDk1MjA1OCwicGF0aCI6Ii8yMzcyODgvNDc2MTE0MTczLWVjMTE0ODRmLTI3YTMtNGVhMi1iNzI0LWZiY2EzNTFiY2NmOC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwODExJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDgxMVQyMjQwNThaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1jNDVhY2RkYjI3NGZjYTNlNWRlMmQ4YmZjMzliMmJiZGM2Yzg0MjAzYjM2OTk1ZmQ1MTg0NjkxNDc3ZTY0NmIzJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.tGg-Q7tAVGRzO6TQCuGNUJzFcgYEaqmVF2z4NKSOlPQ"> <img alt="Screenshot 2025-08-08 at 11 19 58 AM" width="907" height="623" src="https://private-user-images.githubusercontent.com/237288/476114193-7e68a36d-1a21-4c0d-a652-4ec4482f6d28.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTQ5NTIzNTgsIm5iZiI6MTc1NDk1MjA1OCwicGF0aCI6Ii8yMzcyODgvNDc2MTE0MTkzLTdlNjhhMzZkLTFhMjEtNGMwZC1hNjUyLTRlYzQ0ODJmNmQyOC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwODExJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDgxMVQyMjQwNThaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1mY2JhZDJiYmMzZDYzODU1YzJmMDEzMjFiZWMxZTUxN2E5ZjlhYzc3N2UwZWM0YTQzNmIwN2I4ZWFkZTk5ZTY0JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.BWsEAJbeQQaFhYem08w6rH874_6byuOP_NiSGQBBrhQ"> <img alt="Screenshot 2025-08-08 at 11 20 11 AM" width="901" height="914" src="https://private-user-images.githubusercontent.com/237288/476114227-7d3b3904-9aa5-4a5c-ab77-3ab0b7186d7a.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTQ5NTIzNTgsIm5iZiI6MTc1NDk1MjA1OCwicGF0aCI6Ii8yMzcyODgvNDc2MTE0MjI3LTdkM2IzOTA0LTlhYTUtNGE1Yy1hYjc3LTNhYjBiNzE4NmQ3YS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwODExJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDgxMVQyMjQwNThaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT01NjQ5YjUwYmY3M2M1MGVjN2MxMjJlOTJkNjgyZTI0NmZiZDRiZDdhMGQxMjMzYmY2MTYwNWY2YTk2YjcxNTdmJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.XLSOHKj9UUOdf5u2FO5AK3GpcLZ7xyjf0JDp4z292dY">

### Consolidating the feedback from before:

- Under "Distinguishing Different Categoris of Logs" , the log source for session logs is `QuickFix.SessionLogs.<session-id>`, not just `QuickFix.SessionLogs`. e.g. `QuickFix.SessionLogs.FIX.4.4-CLIENT1-EXECUTOR`.
- Under "Log Levels", for completeness I'd add a line saying warnings, errors & exceptions are also logged at their corresponding levels.
- We could mention the new dependency on Microsoft.Extensions.Logging.Abstractions.


Otherwise, the code and docs look good to me. Thanks

jkulubya commented on Examples/Standalone/SerilogLog/README.md:
```diff
@@ -1,35 +1,24 @@
-2. The UnitTests are proof that limits for log size are working.
+This config is meant to demonstrate some features for educational purposes,
+but is probably not something you'd actually *want* to use on a real app
+(for instance, you wouldn't want to roll logs on the minute...).
```
You changed this to hourly
```suggestion
```

gbirchmeier replied:
right, thanks

jkulubya commented on QuickFIXn/Logger/IQuickFixLoggerFactory.cs:
```diff
@@ -0,0 +1,13 @@
+
+/// <summary>
+/// Creates QuickFIX/n-specific ILoggers.
+/// Introduced in v1.15 for Microsoft.Extensions.Logging support.
```
```suggestion
/// Introduced in v1.14 for Microsoft.Extensions.Logging support.
```

gbirchmeier replied:
Good catch.  Jeez, I made this mistake in a few places,  Ooof.

jkulubya commented on QuickFIXn/Session.cs:
```diff
@@ -630,19 +657,19 @@ internal void Next(MessageBuilder msgBuilder)
             catch (FieldNotFoundException e)
             {
-                Log.OnEvent("Rejecting invalid message, field not found: " + e.Message);
+                Log.Log(LogLevel.Information, e, "Rejecting invalid message, field not found: {Message}", e.Message);
```
I think I'd have this as at least at warning level, but that's just preference

gbirchmeier replied:
Agreed. will change.

gbirchmeier commented:
@Rob-Hague wrote:

> There are a few instances where a class is made internal, or a constructor is made internal so the type is no longer publicly instantiable. I assume these are intended, just highlighting to double check

Yes, this was intentional.  IMO, the originators of QF/n made way too much of its innards public, exposing lots of stuff that was never intended for applications to use.  So I'm always looking to reduce the "surface area" of the public interface when it seems appropriate.

gbirchmeier commented:
@jkulubya 

* Thanks for the correction on `<session-id>`, I've updated the table text:

<img width="354" height="241" alt="image" src="https://github.com/user-attachments/assets/5d254155-7559-4a5b-ae68-0ba6128c4fda"/>

.
.
.

* Regarding the new dep, I've added this to the end of the intro:

> **Note:** This change adds the NuGet dependency `Microsoft.Extensions.Logging.Abstractions` to QuickFIX/n.

.

* As for this:

> @jkulubya wrote:
> Under "Log Levels", for completeness I'd add a line saying warnings, errors & exceptions are also logged at their corresponding levels.

I don't see how this conveys any new info that isn't already implied.

However, I did take another pass and revised that subsection.  I fleshed out the text to make it a little more explicit about the other log levels.  What do you think of this?

<img width="754" height="373" alt="image" src="https://github.com/user-attachments/assets/eba48014-409a-4c33-85e4-451418f35b34" />
